stdlib/public/core/StringComparison.swift

```diff
@@ -12,153 +12,244 @@
-      return expecting == _lexicographicalCompare(leftScalar, rightScalar)
+  let scalarDiffIdx = _scalarAlign(utf8Left, diffIdx)
+  _internalInvariant(scalarDiffIdx == _scalarAlign(utf8Right, diffIdx))
+
+  let (leftScalar, leftLen) = _decodeScalar(utf8Left, startingAt: scalarDiffIdx)
+  let (rightScalar, rightLen) = _decodeScalar(
+    utf8Right, startingAt: scalarDiffIdx)
+
+  // Very frequent fast-path: point of binary divergence is a NFC single-scalar
+  // segment. Check that we diverged at the start of a segment, and the next
+  // scalar is both NFC and its own segment.
+  if _fastPath(
+    leftScalar._isNFCStarter && rightScalar._isNFCStarter &&
+    utf8Left.hasNormalizationBoundary(before: scalarDiffIdx &+ leftLen) &&
+    utf8Right.hasNormalizationBoundary(before: scalarDiffIdx &+ rightLen)
+  ) {
+    guard expecting == .less else {
+      // We diverged
+      _internalInvariant(expecting == .equal)
+      return false
     }
+    return _lexicographicalCompare(
+      leftScalar.value, rightScalar.value, expecting: .less)
   }
 
-  // TODO: Back up to start of segment, and slow-path resume from there
+  // Back up to the nearest normalization boundary before doing a slow
+  // normalizing compare.
+  let boundaryIdx = Swift.min(
+    _findBoundary(utf8Left, before: diffIdx),
+    _findBoundary(utf8Right, before: diffIdx))
+  _internalInvariant(boundaryIdx <= diffIdx)
+
+  return _stringCompareSlow(
+    UnsafeBufferPointer(rebasing: utf8Left[boundaryIdx...]),
+    UnsafeBufferPointer(rebasing: utf8Right[boundaryIdx...]),
+    expecting: expecting)
+}
 
-  return _StringGutsSlice(_StringGuts(left, isASCII: false)).compare(with:
-    _StringGutsSlice(_StringGuts(right, isASCII: false)), expecting: expecting)
+@_effects(readonly)
+private func _stringCompareSlow(
+  _ lhs: _StringGuts, _ rhs: _StringGuts, expecting: _StringComparisonResult
+) -> Bool {
+  return _stringCompareSlow(
+    lhs, 0..<lhs.count, rhs, 0..<rhs.count, expecting: expecting)
+}
+
+@_effects(readonly)
+private func _stringCompareSlow(
+  _ lhs: _StringGuts, _ lhsRange: Range<Int>,
+  _ rhs: _StringGuts, _ rhsRange: Range<Int>,
+  expecting: _StringComparisonResult
+) -> Bool {
+  // TODO: Just call the normalizer directly with range
+
+  return _StringGutsSlice(lhs, lhsRange).compare(
+    with: _StringGutsSlice(rhs, rhsRange),
+    expecting: expecting)
+}
+
+@_effects(readonly)
+private func _stringCompareSlow(
+  _ leftUTF8: UnsafeBufferPointer<UInt8>,
+  _ rightUTF8: UnsafeBufferPointer<UInt8>,
+  expecting: _StringComparisonResult
+) -> Bool {
+  // TODO: Just call the normalizer directly
+
+  let left = _StringGutsSlice(_StringGuts(leftUTF8, isASCII: false))
+  let right = _StringGutsSlice(_StringGuts(rightUTF8, isASCII: false))
+  return left.compare(with: right, expecting: expecting)
+}
+
+// Return the point of binary divergence. If they have no binary difference
+// (even if one is longer), returns nil.
+@_effects(readonly)
+private func _findDiffIdx(
```

Comment:
Isn't this just strcspn?

Reply:
Isn't `strcspn` looking for members in a set of characters?

Reply:
Interestingly, almost every bit of documentation I can find is ambiguous on this, but yes, I think you're right. I coulda sworn there was a standard version of this, but I'm blanking on it currently. Anyway, carry on! Sorry about the noise.

Reply:
It's [`std::mismatch`](https://en.cppreference.com/w/cpp/algorithm/mismatch), but you're not going to get that from Swift.